cirq/contrib/two_qubit_gates/math_utils.py

```diff
@@ -0,0 +1,310 @@
+    phi_d = rng.rand(*shape) * np.pi * 2
+    phi_o = rng.rand(*shape) * np.pi * 2
+
+    out = _single_qubit_unitary(theta, phi_d, phi_o)
+
+    if randomize_global_phase:
+        out = np.moveaxis(out, (-2, -1), (0, 1))
+        out *= np.exp(1j * np.pi * 2 * rng.rand(*shape))
+        out = np.moveaxis(out, (0, 1), (-2, -1))
+    return out
+
+
+def vector_kron(first: np.ndarray, second: np.ndarray) -> np.ndarray:
+    """Vectorized implementation of kron for square matrices."""
+    s_0, s_1 = first.shape[-2:], second.shape[-2:]
+    assert s_0[0] == s_0[1]
+    assert s_1[0] == s_1[1]
+    out = np.einsum('...ab,...cd->...acbd', first, second)
+    s_v = out.shape[:-4]
+    return out.reshape(s_v + (s_0[0] * s_1[0],) * 2)
+
+
+# Encode all possible local operations that produce equivalent KAK vectors
+# and which can also be detected by the entanglement fidelity function
+# These operations can be decomposed as s_x^a s_y^b s_z^c n_j p, where
+# s_j denotes a pi/2 shift in index j (a,b,c are 0 or 1), n_j is a pi rotation
+# about the j axis, and p is a permutation of the three indices.
+
+# all permutations of (1,2,3)
+_perms_123 = np.zeros((6, 3, 3), int)
+for ind, perm in enumerate(itertools.permutations((0, 1, 2))):
+    _perms_123[ind, (0, 1, 2), perm] = 1
+
+_negations = np.zeros((4, 3, 3), int)
+_negations[0, (0, 1, 2), (0, 1, 2)] = 1
+_negations[1, (0, 1, 2), (0, 1, 2)] = (1, -1, -1)
+_negations[2, (0, 1, 2), (0, 1, 2)] = (-1, 1, -1)
+_negations[3, (0, 1, 2), (0, 1, 2)] = (-1, -1, 1)
+
+_offsets = np.zeros((8, 3))
+_offsets[1, 0] = np.pi / 2
+_offsets[2, 1] = np.pi / 2
+_offsets[3, 2] = np.pi / 2
+_offsets[4, (1, 2)] = np.pi / 2
+_offsets[5, (0, 2)] = np.pi / 2
+_offsets[6, (0, 1)] = np.pi / 2
+_offsets[7, (0, 1, 2)] = np.pi / 2
+
+
+def _kak_equivalent_vectors(kak_vec) -> np.ndarray:
+    """Generates all KAK vectors equivalent under single qubit unitaries."""
+    # coverage: ignore
+    kak_vec = np.asarray(kak_vec, dtype=float)
+
+    # Apply all permutations, then all negations, then all shifts.
+
+    out = np.einsum('pab,...b->...pa', _perms_123, kak_vec)  # (...,6,3)
+    out = np.einsum('nab,...b->...na', _negations, out)  # (...,6,4,3)
+
+    # (...,8,6,4,3)
+    out = out[..., np.newaxis, :, :, :] + _offsets[:, np.newaxis, np.newaxis, :]
+
+    # Merge indices
+    return np.reshape(out, out.shape[:-4] + (192, 3))
+
+
+def KAK_vector_infidelity(k_vec_a: np.ndarray,
+                          k_vec_b: np.ndarray,
+                          ignore_equivalent_vectors: bool = False
+                         ) -> np.ndarray:
+    """Minimum entanglement infidelity between two KAK vectors. """
+    k_vec_a, k_vec_b = np.asarray(k_vec_a), np.asarray(k_vec_b)
+    if ignore_equivalent_vectors:
+        k_diff = k_vec_a - k_vec_b
+        out = 1 - np.product(np.cos(k_diff), axis=-1)**2
+        out -= np.product(np.sin(k_diff), axis=-1)**2
+        return out
+    # coverage: ignore
```

Comment:
Per your other comment, +1 to adding tests here :)

Reply:
Done. Also added some documentation.

Reply:
Done.